### PR TITLE
Arm64/SVE: Implemented `ConvertToint64` and `ConvertToUInt64`

### PR DESCRIPTION
**Pull request by ebepho**
Contributes to https://github.com/dotnet/runtime/issues/99957

`ConvertToInt64` and `ConvertToUInt64` stress test output:
```
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToInt64_long_float() : 13
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToInt64_long_double() : 13
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToUInt64_ulong_float() : 13
Passed test: _Sve_r::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToUInt64_ulong_double() : 13
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```
```
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToInt64_long_float() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToInt64_long_double() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToUInt64_ulong_float() : 13
Passed test: _Sve_ro::JIT.HardwareIntrinsics.Arm._Sve.Program.Sve_ConvertToUInt64_ulong_double() : 13
===================Running jitstress===================
------------------- {'JitMinOpts': '1'} -------------------
------------------- {'JitStress': '1'} -------------------
------------------- {'JitStress': '2'} -------------------
------------------- {'JitStress': '1', 'TieredCompilation': '1'} -------------------
------------------- {'JitStress': '2', 'TieredCompilation': '1'} -------------------
------------------- {'TailcallStress': '1'} -------------------
------------------- {'ReadyToRun': '0'} -------------------
===================Running jitstressregs===================
------------------- {'JitStressRegs': '1'} -------------------
------------------- {'JitStressRegs': '2'} -------------------
------------------- {'JitStressRegs': '3'} -------------------
------------------- {'JitStressRegs': '4'} -------------------
------------------- {'JitStressRegs': '8'} -------------------
------------------- {'JitStressRegs': '0x10'} -------------------
------------------- {'JitStressRegs': '0x80'} -------------------
------------------- {'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStressRegs': '0x2000'} -------------------
===================Running jitstress2-jitstressregs===================
------------------- {'JitStress': '2', 'JitStressRegs': '1'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '2'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '3'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '4'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '8'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x10'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x80'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x1000'} -------------------
------------------- {'JitStress': '2', 'JitStressRegs': '0x2000'} -------------------
```

cc @dotnet/arm64-contrib

### CODE REVIEW

File: src/coreclr/jit/hwintrinsic.cpp [hunk 1]
```diff
@@ -1731,6 +1731,8 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
                 {
                     case NI_Sve_ConvertToInt32:
                     case NI_Sve_ConvertToUInt32:
+                    case NI_Sve_ConvertToInt64:
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsic.cpp:1734]: I expected to have entries for `ConvertToSingle` and `ConvertToDouble` here as well. Did you confirm this works for scenario where `maskSize != operSize` in lowering?

File: src/coreclr/jit/lowerarmarch.cpp [hunk 2]
```diff
@@ -3388,8 +3388,10 @@ void Lowering::ContainCheckHWIntrinsic(GenTreeHWIntrinsic* node)
 
                             // For now, make sure that we get here only for intrinsics that we are
                             // sure about to rely on auxiliary type's size.
-                            assert((embOp->GetHWIntrinsicId() == NI_Sve_ConvertToInt32) ||
-                                   (embOp->GetHWIntrinsicId() == NI_Sve_ConvertToUInt32));
+                            assert((embOp->GetHWIntrinsicId() == NI_Sve_ConvertToInt32)  ||
+                                   (embOp->GetHWIntrinsicId() == NI_Sve_ConvertToUInt32) ||
+                                   (embOp->GetHWIntrinsicId() == NI_Sve_ConvertToInt64)  ||
```
Comment (kunalspathak) [inline src/coreclr/jit/lowerarmarch.cpp:3393]: likewise here.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 3]
```diff
@@ -492,12 +492,19 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                    insOpts embOpt = opt;
                     switch (intrinEmbMask.id)
                     {
                         case NI_Sve_ConvertToInt32:
                         case NI_Sve_ConvertToUInt32:
                         {
-                            opt = intrinEmbMask.baseType == TYP_DOUBLE ? INS_OPTS_D_TO_S : INS_OPTS_SCALABLE_S;
+                            embOpt = intrinEmbMask.baseType == TYP_DOUBLE ? INS_OPTS_D_TO_S : INS_OPTS_SCALABLE_S;
+                            break;
+                        }
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:503]: This also needs handling of `ConvertToSingle` 

```c++
embOpt = intrinEmbMask.baseType == TYP_DOUBLE ? INS_OPTS_D_TO_S : INS_OPTS_SCALABLE_S;
```

and `ConvertToDouble`:

```c++
embOpt = intrinEmbMask.baseType == TYP_FLOAT ? INS_OPTS_S_TO_D : INS_OPTS_SCALABLE_D;
```

which means you can combine `ConvertToSingle` with `Convert*32` and `ConvertToDouble` with `Convert*64`.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 4]
```diff
@@ -78,9 +78,9 @@
                 }
             }";
 
-const string SimpleVecOpTest_ValidationLogicForCndSelForNarrowing = @"for (var i = 0; i < Op1ElementCount; i++)
+const string SimpleVecOpTest_ValidationLogicForCndSelNarrowing = @"for (var i = 0; i < Op1ElementCount; i++)
```
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:81]: why this change?

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 5]
```diff
@@ -3084,10 +3084,16 @@
...
-    ("SveSimpleVecOpDifferentRetTypeTest.template",        new Dictionary<string, string> { ["TestName"] = "Sve_ConvertToUInt32_uint_double",                                                                       ["Isa"] = "Sve",           ["LoadIsa"] = "Sve",     ["Method"] = "ConvertToUInt32",                                                      ["RetVectorType"] = "Vector",    ["RetBaseType"] = "UInt32",  ["Op1VectorType"] = "Vector",    ["Op1BaseType"] = "Double",                                                                                                                           ["LargestVectorSize"] = "64", ["NextValueOp1"] = "TestLibrary.Generator.GetDouble()",                                                                                                                                                      ["ValidateIterResult"] = "Helpers.ConvertDoubleToUInt32(firstOp[i]) != result[i]", ["GetIterResult"] = "Helpers.ConvertDoubleToUInt32(leftOp[i])"}),
+
```
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:3087]: can you delete the 2 empty lines?

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 6]
```diff
@@ -177,7 +177,7 @@
...
     ("_SveUnaryOpTestTemplate.template",                 "SveSimpleVecOpTest.template",                 new Dictionary<string, string> { ["TemplateName"] = "Simple",     ["TemplateValidationLogic"] = SimpleVecOpTest_ValidationLogic,             ["TemplateValidationLogicForCndSel"] = SimpleVecOpTest_ValidationLogicForCndSel }),
-    ("_SveUnaryOpDifferentRetTypeTestTemplate.template", "SveSimpleVecOpDifferentRetTypeTest.template", new Dictionary<string, string> { ["TemplateName"] = "Simple",     ["TemplateValidationLogic"] = SimpleVecOpTest_ValidationLogicForNarrowing, ["TemplateValidationLogicForCndSel"] = SimpleVecOpTest_ValidationLogicForCndSelForNarrowing }),
+    ("_SveUnaryOpDifferentRetTypeTestTemplate.template", "SveSimpleVecOpDifferentRetTypeTest.template", new Dictionary<string, string> { ["TemplateName"] = "Simple",     ["TemplateValidationLogic"] = SimpleVecOpTest_ValidationLogicForNarrowing, ["TemplateValidationLogicForCndSel"] = SimpleVecOpTest_ValidationLogicForCndSelNarrowing }),
```
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:180]: likewise here.

**Review by kunalspathak — CHANGES_REQUESTED** [review]
Added some comments around test. code changes looks good.

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs (hunk 4 above)
Comment (ebepho) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:81]: I reverted this change -- shouldn't have been in this PR

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs [hunk 7]
```diff
@@ -80,7 +80,7 @@
 
 const string SimpleVecOpTest_ValidationLogicForCndSelForNarrowing = @"for (var i = 0; i < Op1ElementCount; i++)
             {
-                {Op1BaseType} iterResult = (mask[i] != 0) ? {GetIterResult} : falseVal[i];
+                {RetBaseType} iterResult = (mask[i] != 0) ? {GetIterResult} : falseVal[i];
```
Comment (kunalspathak) [inline src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:83]: Is this needed too?

Comment (kunalspathak) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs:83]:
```suggestion
                {Op1BaseType} iterResult = (mask[i] != 0) ? {GetIterResult} : falseVal[i];
```

Comment (kunalspathak) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs]: fixed it.

**Review by kunalspathak — APPROVED** [review]
LGTM

File: src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs (hunk 7 above)
Comment (ebepho) [reply src/tests/Common/GenerateHWIntrinsicTests/GenerateHWIntrinsicTests_Arm.cs]: I'm going to test this version one more time. Not sure why these changes even appeared, and I think they were supposed to be part of the `ConvertToInt32` commit. These changes may be necessary for the tests to run properly.
